source/common/router/router.cc

```diff
@@ -778,5 +763,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
```

Comment:
I think this is risky and needs a lot of thinking. Cleanup may/will occur in the normal reset/end_stream case at which point we need to consider the stream dead. It's possible this is safe because of your new guards in the codec helper, but I think this needs careful thought and testing.